lambdaclass/mirra_backend
# [GH-640] Endpoint authentication

## Motivation

Closes #640 

## Summary of changes

Uff... So not much, but at the same time a lot here. All changes here are based on this [swimlane](https://swimlanes.io/u/EdQr-F_ZG)

![swimlanes-aa55333e05412682407e5e7976eed6af](https://github.com/lambdaclass/mirra_backend/assets/3791966/02558203-80f1-430c-92ec-43ea166ccf34)


- Added a module to handle JWTs emitted by `gateway` app
  - [x] Add `JWT_PRIVATE_KEY` to servers
- In `arena` expect a JWT when connecting to the websockets (`SocketHandler` and `GameSocketHandler`), verify it, and extract `user_id` from it
  - [x] #678 ~Create issue to address TODO due to bots been broken for the authentication and probably load testing is broken as well~
- Replaced associations to `google_user` with regular `user`
- Created endpoint in `gateway` to expose the public key clients (other apps) should use to verify the JWTs emitted
- Added endpoint to create a guest user
  - [x] #679  ~reate issue to address security gap in such an endpoint, right now anyone that can reach it will be able to create users~ 
- Added `Guest login` button in `game_client` to allow guest login, google login is still possible, but Google OAuth creds are required
- [x] Route to refresh tokens
- [x] Follow up issue #695  

## How to test it?

Testing wise the main thing is to play with all clients (Unity, apps, and game_client) and make sure everything keeps working. Besides that you should be able to see the match results been stored in the DB

For Unity/apps test with https://github.com/lambdaclass/champions_of_mirra/pull/1803

## Checklist
